src/modules/referral/services/service.ts

```diff
@@ -71,4 +75,47 @@ export class ReferralService {
 
     return { referralRate: 0.4, tier: 1 };
   }
+
+  public subtractFunctionArgsFromCallData(data: string) {
+    const coder = new ethers.utils.AbiCoder();
+    // strip hex method identifier
+    const dataNoMethod = ethers.utils.hexDataSlice(data, 4);
+    // keep method hex identifier
+    const methodHex = data.replace(dataNoMethod.replace("0x", ""), "");
+    const decodedData = coder.decode(
+      ["address", "address", "uint256", "uint256", "uint64", "uint32"],
+      ethers.utils.hexDataSlice(data, 4),
+    );
+    const encoded = coder.encode(["address", "address", "uint256", "uint256", "uint64", "uint32"], decodedData);
+    const fullEncoded = methodHex + encoded.replace("0x", "");
+    return data.replace(fullEncoded, "");
+  }
+
+  public extractReferralAddressUsingDelimiter(data: string) {
+    const referralData = this.subtractFunctionArgsFromCallData(data);
+
+    if (referralData.indexOf(REFERRAL_ADDRESS_DELIMITER) !== -1) {
+      const addressIndex = referralData.indexOf(REFERRAL_ADDRESS_DELIMITER) + REFERRAL_ADDRESS_DELIMITER.length;
+      const potentialAddress = referralData.slice(addressIndex, addressIndex + 40);
```

Comment:
Should we do this instead to also weed out cases where there are _more than_ 40 additional characters?

```suggestion
      const potentialAddress = referralData.slice(addressIndex);
```

Reply:
@mrice32  I took into consideration the fact that in some cases the referral address can be found at other positions than the end of the input string 🤷‍♂️